packages/expo-app-auth/android/src/main/java/expo/modules/appauth/AppAuthModule.java

```diff
@@ -0,0 +1,439 @@
+package expo.modules.appauth;
+
+import android.app.Activity;
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
+import android.net.Uri;
+import android.os.Bundle;
+import android.support.annotation.Nullable;
+
+import net.openid.appauth.AppAuthConfiguration;
+import net.openid.appauth.AuthorizationException;
+import net.openid.appauth.AuthorizationRequest;
+import net.openid.appauth.AuthorizationResponse;
+import net.openid.appauth.AuthorizationService;
+import net.openid.appauth.AuthorizationServiceConfiguration;
+import net.openid.appauth.ClientAuthentication;
+import net.openid.appauth.ClientSecretBasic;
+import net.openid.appauth.ResponseTypeValues;
+import net.openid.appauth.TokenRequest;
+import net.openid.appauth.TokenResponse;
+import net.openid.appauth.connectivity.ConnectionBuilder;
+import net.openid.appauth.connectivity.DefaultConnectionBuilder;
+
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Locale;
+import java.util.Map;
+import java.util.TimeZone;
+
+import de.greenrobot.event.EventBus;
+import expo.core.ExportedModule;
+import expo.core.ModuleRegistry;
+import expo.core.Promise;
+import expo.core.interfaces.ActivityEventListener;
+import expo.core.interfaces.ActivityProvider;
+import expo.core.interfaces.ExpoMethod;
+import expo.core.interfaces.ModuleRegistryConsumer;
+import expo.core.interfaces.services.UIManager;
+import expo.interfaces.constants.ConstantsInterface;
+
+public class AppAuthModule extends ExportedModule implements ModuleRegistryConsumer {
+    private static final String TAG = "ExpoAppAuth";
+    private static final String ERROR_TAG = "E_APP_AUTH";
+    private ModuleRegistry mModuleRegistry;
+    private static final String MANIFEST_URL_KEY = "experienceUrl";
+
+    private AuthTask authTask = new AuthTask();
+    private Boolean canMakeInsecureRequests;
+    private Map<String, String> additionalParametersMap;
+    private String clientSecret;
+
+    public AppAuthModule(Context context) {
+        super(context);
+    }
+
+    @Override
+    public void setModuleRegistry(ModuleRegistry moduleRegistry) {
+        mModuleRegistry = moduleRegistry;
+    }
+
+    @Override
+    public String getName() {
+        return TAG;
+    }
+
+    private Activity getCurrentActivity() {
+        if (mModuleRegistry != null) {
+            ActivityProvider activityProvider = mModuleRegistry.getModule(ActivityProvider.class);
+            return activityProvider.getCurrentActivity();
+        }
+        return null;
+    }
+
+    private Map<String, String> castObjectsToStrings(Map<String, Object> map) {
+        Map<String, String> newMap= new HashMap<>();
+        for(String strKey: map.keySet()) {
+            newMap.put(strKey, String.valueOf(map.get(strKey)));
+        }
+        return newMap;
+    }
+
+    @ExpoMethod
+    public void executeAsync(
```

Comment:
this method seems especially large and unwieldy -- i wonder if it might help to split out the branches of `if (serviceConfiguration != null)` out into their own methods to clean things up a bit